File: tests/cpp/src/TestBasicResponder.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include <cameo/api/cameo.h>
#include <iostream>

using namespace std;
using namespace cameo;

int main(int argc, char *argv[]) {

	This::init(argc, argv);

	string applicationName;

	int numberOfTimes = 1;

	if (argc > 2) {
		applicationName = argv[1];
		cout << "Responder application is " + applicationName << endl;

		if (argc > 3) {
			numberOfTimes = stoi(argv[2]);
		}
	}
	else {
		cerr << "Arguments: [application name] <number of times>" << endl;
		return -1;
	}

	int options = 0;
	bool useProxy = false;
	string endpoint = "tcp://localhost:11000";
	if (argc > 4) {
		useProxy = (string(argv[3]) == "true");
	}
	if (useProxy) {
		options |= option::USE_PROXY;
		endpoint = "tcp://localhost:12000";
	}

	unique_ptr<Server> server = Server::create(endpoint, options);
	server->init();

	// Loop the number of times.
	for (int i = 0; i < numberOfTimes; ++i) {

		// Args.
		vector<string> args{(useProxy ? "true" : "false")};

		// Start the application.
		unique_ptr<App> responderApplication = server->start(applicationName, args);

		cout << "Started application " << *responderApplication << " with state " << toString(responderApplication->getState()) << endl;

		// Create a subscriber to the application applicationName
		unique_ptr<coms::Requester> requester = coms::Requester::create(*responderApplication, "responder");
		requester->setCheckApp(true);

		cout << "Requester ready ? " << requester->isReady() << endl;
		requester->init();
		cout << "Requester ready ? " << requester->isReady() << endl;

		// Try a second call.
		requester->init();

		cout << "Created requester " << *requester << endl;

		if (!requester) {
			cout << "Requester error" << endl;
			return -1;
		}

		cout << "Application " << *responderApplication << " has state " << toString(responderApplication->getState()) << endl;

		// Send a simple message.
		requester->send("request");

		optional<string> response = requester->receive();
		cout << "Response is " << response.value() << endl;

		response = requester->receive();
		cout << "Response 2 is " << response.value() << endl;

		// Send a two-parts message.
		requester->sendTwoParts("first", "second");

		response = requester->receive();
		cout << "Response is " << response.value() << endl;

		// Send a simple message but do not receive the response immediately.
		requester->send("request");

		cout << "Wait so that the responder has replied" << endl;
		this_thread::sleep_for(chrono::seconds(1));

		response = requester->receive();
		cout << "Response is " << response.value() << endl;

		// Send a new simple message.
		requester->send("request after timeout");
		requester->setTimeout(200);

		response = requester->receive();

		if (response.has_value()) {
			cout << "Response is " << response.value() << endl;
		}
		else {
			if (requester->hasTimedout()) {
				cout << "Timeout" << endl;
			}
			else {
				cout << "No response" << endl;
			}
		}

		// The requester needs to resync after a timeout.
		// If the server does not respond within the configured timeout, an error occurs.
		requester->send("request after timeout");
		if (requester->hasTimedout()) {
			cout << "Timeout while resyncing" << endl;
		}

		cout << "Wait so that the server is able to respond" << endl;
		this_thread::sleep_for(chrono::seconds(1));

		// Resend the request.
		requester->send("request after timeout");
		if (!requester->hasTimedout()) {
			cout << "No timeout while sending" << endl;
		}

		response = requester->receive();
		cout << "Response is " << response.value() << endl;

		// Cancel the requester.
		requester->cancel();
		requester->send("request after cancel");
		response = requester->receive();

		if (response.has_value()) {
			cout << "Response is " << response.value() << endl;
		}
		else {
			if (requester->hasTimedout()) {
				cout << "Timeout" << endl;
			}
			else if (requester->isCanceled()) {
				cout << "Canceled" << endl;
			}
		}

		// Re-init the requester has no effect.
		requester->init();
		requester->send("2nd request after cancel");
		response = requester->receive();

		if (response.has_value()) {
			cout << "Response is " << response.value() << endl;
		}
		else {
			if (requester->hasTimedout()) {
				cout << "Timeout" << endl;
			}
			else if (requester->isCanceled()) {
				cout << "Canceled" << endl;
			}
		}

		// Wait for the end of the application.
		state::Value state = responderApplication->waitFor();

		cout << "Responder application terminated with state " << toString(state) << endl;
		cout << "Finished the application" << endl;
	}

	return 0;
}

File: tests/python/src/testbasicresponder.py
import sys
import time
import cameopy

this = cameopy.This
this.init(sys.argv)

applicationName = ""
numberOfTimes = 1 

if len(sys.argv) >= 3:
    applicationName = sys.argv[1]
    
if len(sys.argv) >= 4:
    numberOfTimes = int(sys.argv[2])

useProxy = False if len(sys.argv) < 5 else (sys.argv[3] == "true") 

options = 0
endpoint = "tcp://localhost:11000";
if useProxy:
    options |= cameopy.option.USE_PROXY
    endpoint = "tcp://localhost:12000";

server = cameopy.Server.create(endpoint, options)
server.init()

print("Responder application is", applicationName)

for i in range(numberOfTimes):
    
    args = ["true" if useProxy else "false"]
    app = server.start(applicationName, args)
    print("Started application", applicationName)

    requester = cameopy.coms.Requester.create(app, "responder")
    requester.setCheckApp(True)
    requester.init()
    
    print("Created requester", requester)
    
    # Send a simple message.
    requester.send("request")
    
    response = requester.receiveString()
    print("Response is", response)
    response = requester.receiveString()
    print("Response 2 is", response)
    
    # Send a two-parts message.
    requester.sendTwoParts("first", "second")
    
    response = requester.receiveString()
    print("Response is", response)
    
    # Send a simple message but do not receive the response immediately.
    requester.send("request");
    
    print("Wait so that the responder has replied")
    time.sleep(1)
    
    response = requester.receiveString()
    print("Response is", response)
    
    # Send a new simple message.
    requester.send("request after wait")
    requester.setTimeout(200)

    response = requester.receiveString()
    
    if not response is None:
        print("Response is", response);
    elif requester.hasTimedout():
        print("Timeout")    
    else:
        print("No response")    

    # The requester needs to resync after a timeout.
    # If the server does not respond within the configured timeout, an error occurs.
    requester.send("request after timeout")
    if requester.hasTimedout():
        print("Timeout while resyncing")
    
    print("Wait so that the server is able to respond")            
    time.sleep(1)
        
    # Resend the request.
    requester.send("request after timeout")
    if not requester.hasTimedout():
        print("No timeout while sending")
    
    response = requester.receiveString();
    print("Response is", response)
    
    # Wait for the application.
    app.waitFor()
    
    requester.terminate()
    
    print("Finished the application")


    